### DISCUSSION TITLE
Revert commit on Website

### DISCUSSION THREAD

**Discussion by BtucoGit** [community/community#60023] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Hello there!
I'm beginner on GitHub, and can't resolve one problem - on desktop version i can "Revert changes in commit" in history without problems. But i can't find any way to revert commits on Website. I need to use the website and not the git console. A search on the web yielded no results - everywhere is used console. It is possible to use only website?

**Comment by GhostOf0days** [comment]
Hey. If you don't want to revert a commit using the console, the easiest way is to go to [Github Desktop](https://desktop.github.com/) (which is UI-based rather than terminal), set current repo to the specific repo, click on history, and right click on a commit. You will now see a button that says "Revert changes in commit."
![image](https://github.com/community/community/assets/70134418/94dfca81-3b13-4ace-a05a-01cda3d8d195)
Click it then click push origin at the top. 
![image](https://github.com/community/community/assets/70134418/631ae52c-b0e0-4ddb-9c89-53f22b10e936)

**Reply by GhostOf0days** [reply]
Regardless, please do not avoid git for terminal. It is extremely useful to you as a dev. Avoiding it will only hurt you. It's easy to setup: https://docs.github.com/en/get-started/quickstart/set-up-git. To revert a commit you can ```git reset --hard<commit-hash>``` then ```git push origin <branch-name>```, where <branch-name> would likely be main for you as I assume you don't know much about branching yet. 
Commit hash is the string that appears to the right of a commit when you click the following:
![image](https://github.com/community/community/assets/70134418/c4aed7b2-5b26-4783-9df8-278de1f97ba8).

You can also do ```git reset --hard HEAD~number```, where number is how many commits you want to reset to (so HEAD~1 would revert to your last commit). After that, do push again as specified before. You can also do soft instead of hard. Soft keeps changes after that commit you reset to, so you can make modifications, but you are now at the commit you reset to. Hard resets the staging area (where you see  to the commit (basically where you see the commits below) and starts you off at a clean slate as if you made no changes since the commit you reset to. 
![image](https://github.com/community/community/assets/70134418/7b9c55df-a727-4de0-837e-0fda1d5233f9)

You can also force push by adding ```--force``` after the word "push" in ```git push origin <branch-name>``` (again, for you, branch name is likely main). This forces the remote repo (on Github website) to have the changes you made with the force push even if you didn't pull (bring in changes that you had on the remote repo but nor locally). So if you have 3 commits (changes) that you didn't pull to the local version on your computer, they would be overwritten with a force push.

**Reply by BtucoGit** [reply]
Yes, my dear ones, I know about the possibilities and benefits of the console, as well as about the possibility of revert commit from the application to PC. I'm interested in revert commit through the Website, because there is not always access not only to the console, but also to the computer as a whole. It is possible?

**Reply by GhostOf0days** [reply, marked as answer]
Not as of this moment. Github doesn't have a web API call to do it, so it's not possible. [See their API documentation](https://docs.github.com/en/rest?apiVersion=2022-11-28) to check what's possible.

**Reply by GhostOf0days** [reply]
If the API call was available, you would be able to do it.

**Reply by GhostOf0days** [reply]
There is an exception, however, in that you can revert commits for merged pull requests on the website, so if you make your commits as pull requests, I guess that's a bypass to reverting commits on the website.

**Reply by BtucoGit** [reply]
Many thanks! That one of reasons to learn console more tighter interaction

**Reply by kdehl16-web** [reply]
GitHub*

**Reply by eden-amari** [reply]
You're a lifesaver, thank you!!

**Comment by Turbulentarius** [comment]


**Reply by dhust** [reply]
We actually can't undo a commit on the website? That's crazy! How is an "undo" operation not possible?

I ask because I accidently made an additional commit using Github Desktop. To go back I right-clicked and chose `Revert changes in commit` but it gave me an error. I'm like "Ok, I'll just go to the website and do it there." Nope. :( Once year I go back to trying Github and it ticks me off within a day. How is this the norm? This is probably the only service/website I hate using. Rant over.

**Reply by levino** [reply]
Compared to what? Which service are you using instead of GitHub?

**Comment by aly-fly** [comment]
Thanks for the tips. 
TLDR - here are the steps to revert:
- open the repo on website, click the icon with the `clock and <nn> Commits`
- copy the full hash of the time you want revert to (upper position is later in time)
- use command: `git reset --hard <long hash number copied>`
- review that source files are with contents as you want them
- use command `git push origin main --force` to copy current state on the computer to the server
- refresh the website and latest unwanted commit(s) will disappear.
- make a happy face :)

**Reply by Siegfriedk** [reply]
This is not how you revert a git commit and not a good suggestion. This is how you reset the github repository to a previous version. If anyone did something while you are 'resetting' the changes from the other person will disappear.

There is a real and proper way to do so:

You use git revert https://git-scm.com/docs/git-revert

**Reply by aly-fly** [reply]
This is exactly what I wanted to do - delete all changes without a trace - if you accidentally push a password or other sensitive data, you want to completely wipe it, so that it is not accessible in the history. Without deleting and re-creating the whole repository.

**Reply by Sicaine** [reply]
Its just not stated in your comment and the implications of git conflicts (when working with others) was also not mentioned. Nonetheless, secrets/passwords are compromised as soon as you commited them. This only makes sense for sensitive data

**Reply by Siegfriedk** [reply]
> > This is not how you revert a git commit and not a good suggestion. This is how you reset the github repository to a previous version. If anyone did something while you are 'resetting' the changes from the other person will disappear.
> > There is a real and proper way to do so:
> > You use git revert https://git-scm.com/docs/git-revert
> 
> this is not a solution u fucking retard

You need to start learning to read and comprehend instead of being a dick.

Again, it makes a difference if you work alone on your github repo or with a team and it makes a difference if it is a password or sensitive data. A pushed secret is a secret you need to reset on the system were it came from as its now officially leaked. There are plenty of systems out there who scrap GitHub constantly. 

Sensitive Data its different, then you still need to explain the whole process which means, that if you part of a team, you need to coordindate with the others if you want to do a hard reset.

**Reply by alysdexia** [reply]
> > > This is not how you revert a git commit and not a good suggestion. This is how you reset the github repository to a previous version. If anyone did something while you are 'resetting' the changes from the other person will disappear.
> > > There is a real and proper way to do so:
> > > You use git revert https://git-scm.com/docs/git-revert
> > 
> > 
> > this is not a solution u fucking retard
> 
> You need to start learning to read and comprehend instead of being a dick.
> 
> Again, it makes a difference if you work alone on your github repo or with a team and it makes a difference if it is a password or sensitive data. A pushed secret is a secret you need to reset on the system were it came from as its now officially leaked. There are plenty of systems out there who scrap GitHub constantly.
> 
> Sensitive Data its different, then you still need to explain the whole process which means, that if you part of a team, you need to coordindate with the others if you want to do a hard reset.

“the other person will disappear”, “were”, “its”, “who scrap”, “its”, “you part”, “coordindate”, all these diction abominations yet “You need to start learning to read and comprehend”.

**Comment by MadeleineMason** [comment]
GitHub’s website lacks a revert button, so you’ll need to manually restore files, had to do the same on [Pearl Lemon Web](https://pearllemonweb.com/)’s repo once.

**Comment by AMDphreak** [comment]
This is incredibly stupid. GitHub, why must us simple folk constantly have to teach you what is good UI/UX?

**Comment by DonaldDuck313** [comment]
My use case is that I have a huge repository that takes hours to download and I just want to revert a commit in it. I can't use the terminal for that because then I would have to run git clone which takes hours, so I tried to install GitHub Desktop, but that didn't help because I would have had to clone the repository there also. It seems like the quickest in my case is to just manually copy-paste the old code into a new commit on the website.

**Reply by Sicaine** [reply]
Besides that i find it also very unfortunate, that GitHub doesn't support it on the UI, you should be able to do a shallow clone (https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone/) with a small depth size to fix it without having to download a lot.

On a side note: If your repository really takes hours to download, you might have some serious issues in your repo. Do you checkin a lot of files?

**Reply by DonaldDuck313** [reply]
It's not my repo, it's https://github.com/DefinitelyTyped/DefinitelyTyped. They have TypeScript type definitions for most JavaScript libraries so their repo is huge, which is why it's so slow to download.

**Comment by levino** [comment]
FWIW: As a workaround, one can use github copilot coding agent and ask it to "revert commit ...". It will create a PR with a revert commit (in most cases, I assume).
<img width="774" height="475" alt="Screenshot from 2025-09-08 12-10-18" src="https://github.com/user-attachments/assets/5413ca59-db39-4901-be25-31ffd407b306" />

I find this actually pretty good. For me it would be a bit overkill to extend the already very cluttered interface of GitHub with such functionality. Also it is often not so easy to revert a commit without encountering conflicts. This is what copilot should be able to solve.

The future is now. We need no more buttons. Just tell the agent in plain English what you want to be done.

**Reply by ichalis** [reply]
Thanks for this feedback! Who would've thought rolling back a change would be so stupidly difficult/unintuitive! Isnt this one of the whole points of github? Every time I try to learn github, i just end up getting annoyed...

**Reply by KaKi87** [reply]
> it would be a bit overkill to extend the already very cluttered interface of GitHub with such functionality

But it wouldn't be **a lot** overkill to compensate with running resource-intensive LLM inference just to execute an action that could be a simple deterministic button ?

**Comment by BonifacioCalindoro** [comment]
October 2025 and we still can't revert commits from the WebUI? Cmon github

**Reply by intromaster** [reply]
Dec 2025, still cant revert commits! crazy

**Reply by hiagocosta** [reply]
> Dec 2025, still cant revert commits! crazy

Mar 2026 and still can't revert commits, crazier

**Reply by erkr** [reply]
APRIL 😢

**Reply by astridha** [reply]
😒

**Reply by collectivmedia** [reply]
Checking in for May 2026. *cries in new developer*

**Reply by keith-nielsen** [reply]
>Is GitHub now owned by Microsoft?
GitHub, headquartered in San Francisco, is operated by Github, Inc., a subsidiary of Microsoft since 2018. San Francisco, California, U.S.

The only "solution" is to use Copilot to fix the thing that is literally the very first thing a new user absolutely needs to do when they mess up.

This is working as intended people... Hegel's Dielectric Cycle writ large.

Create a thing that seems useful, but has a glaring/illogical gap, to fill the gap you must juice up the thing that they need juiced, stock-dependent metric gets juiced to hire more devs that will pump the HDC.

Pure evil.

**Comment by wayne-schwebel-cs** [comment]
Hi, I know this is an old discussion but hoping someone might have some updated info here. Our situation is a bit more nuanced.

One of the team raised a PR against our main branch instead of a release branch accidentally and no one noticed until it was too late. Consequently a massive amount of changes were prematurely merged to it. 

What we want to do it RESET the head back to just before that commit. Due to company enforced policies, we cannot commit directly to main without a PR (kind of sensible). 

What many people here seem to be missing is that a REVERT can often NOT be what you want to do. In my experience, way more often than not, that is the case.

In this case, if we were to revert the commit, the following would happen:

* Git will create a new commit with the same number of changes but in reverse i.e. setting all the files back to what they were before
* We would indeed be able to raise a PR and merge those changes into main (seems good so far)
* When the feature is actually finished and ready to merge to main, git would see that the majority of changes have already been merged, then undone and merged again and practically none of the original, correct changes will be in the merge.
* We would have to then manually apply ALL of the changes again and commit them as a brand new commit.

So far, the team has been blocked from interacting with the main branch at all, so no issue with anyone having local branches with the errant commit in them.

It would be ideal if GitHub had some function to be able to do this in the site itself as it is NOT the same scenario as committing directly to the main branch that we are trying to block NOR is it the same as doing a revert which is ABSOLUTELY NOT what we want to do here.
